modules/call-dispatch/src/lib.rs

```diff
@@ -0,0 +1,348 @@
+	///
+	/// 1) `InstanceId` of deployed bridge module => if you want all messages coming
+	/// through this bridge to be dispatched using the same origin;
+	/// 2) `InstanceId` + `LaneId` => if you want all messages coming through given
+	/// lane of given bridge to be dispatched using the same origin
+	type MessageOrigin: Parameter;
+	/// Id of the message. Whenever message is passed to the dispatch module, it emits
+	/// event with this id + dispatch result.
+	type MessageId: Parameter;
+	/// The overarching dispatch call type.
+	type Call: Parameter + GetDispatchInfo + Dispatchable<Origin = <Self as frame_system::Trait>::Origin>;
+}
+
+decl_event!(
+	pub enum Event<T> where
+		<T as Trait>::MessageOrigin,
+		<T as Trait>::MessageId,
+	{
+		/// Message has been rejected by dispatcher because of spec version mismatch.
+		MessageVersionSpecMismatch(MessageOrigin, MessageId),
+		/// Message has been rejected by dispatcher because of weight mismatch.
+		MessageWeightMismatch(MessageOrigin, MessageId),
+		/// Message has been dispatched with given result.
+		MessageDispatched(MessageOrigin, MessageId, DispatchResult),
+	}
+);
+
+decl_module! {
+	/// Call Dispatch FRAME Pallet.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Deposit one of this module's events by using the default implementation.
+		fn deposit_event() = default;
+	}
+}
+
+impl<T: Trait> MessageDispatch<T::MessageOrigin, T::MessageId> for Module<T>
+where
+	<<T as Trait>::Call as Dispatchable>::PostInfo: sp_std::fmt::Debug,
+{
+	type Message = (SpecVersion, Weight, <T as Trait>::Call);
+
+	fn dispatch(origin: T::MessageOrigin, id: T::MessageId, message: Self::Message) {
+		let (spec_version, weight, call) = message;
+
+		// verify spec version
+		// (we want it to be the same, because otherwise we may decode Call improperly)
+		let expected_version = <T as frame_system::Trait>::Version::get().spec_version;
+		if spec_version != expected_version {
+			frame_support::debug::trace!(
+				"Message {:?}/{:?}: spec_version mismatch. Expected {:?}, got {:?}",
+				origin,
+				id,
+				expected_version,
+				spec_version,
+			);
+			Self::deposit_event(Event::<T>::MessageVersionSpecMismatch(origin, id));
+			return;
+		}
+
+		// verify weight
+		// (we want passed weight to be at least equal to pre-dispatch weight of the call
+		// because otherwise Calls may be dispatched at lower price)
+		let expected_weight = call.get_dispatch_info().weight;
+		if weight < expected_weight {
+			frame_support::debug::trace!(
+				"Message {:?}/{:?}: passed weight is too low. Expected at least {:?}, got {:?}",
+				origin,
+				id,
+				expected_weight,
+				weight,
+			);
+			Self::deposit_event(Event::<T>::MessageWeightMismatch(origin, id));
+			return;
+		}
+
+		// finally dispatch message
+		let origin_account = Self::bridge_account_id(&origin);
+		let dispatch_result = call.dispatch(frame_system::RawOrigin::Signed(origin_account).into());
```

Comment:
We will most likely need to keep track of consumed `weight` at some level. AFAIU The stack looks like this:
1. Delivery (external call)
2. Scheduler
3. Dispatch (internal `call.dispatch`)

Substrate requires us to provide `weight` upper bound at (1). The simplest calculation would be just a `Min(block_weight_left, SUM(weight_of_each_message))`, a better solution would delegate to scheduler, which could restrict this even further. Then during the actual dispatch, we should keep track of actually consumed weight to: a) let the scheduler process more calls if possible and needed, b) let the relayer pay smaller fees for (1), and c) better utilize available block weight.

Reply:
Changed return type to `Weight`. I had to 'fix' `Dispatchable::PostInfo` to be `frame_support::dispatch::PostDispatchInfo`. Otherwise it seems impossible to get actual weight.

Some extended context: I expect scheduler to only rely to remaining weight of current block (i.e. `frame_system::BlockWeight`) and maximal weight of current block. So imo this change isn't required - it is dispatcher who needs to update `frame_system::BlockWeight` (it is missing from current PR, but most probably will be changed in follow-up PRs).